arrow-parens ("as-needed") doesn't take async into account

**Tell us about your environment**

* **ESLint Version:** 3.19.0 (with the latest babel-eslint)
* **Node Version:** 7.9.0
* **npm Version:** 4.2.0

**What parser (default, Babel-ESLint, etc.) are you using?** babel-eslint (as eslint doesn't support async

**Please show your full configuration:**

final test with the following, still breaks:
```
{
  "root": true,
  "parser": "babel-eslint",
  "parserOptions": {
    "ecmaVersion": 7,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true,
      "impliedStrict": true,
      "experimentalObjectRestSpread": true
    }
  },
  "env": {
    "browser": true,
    "mocha": true,
    "node": true,
    "es6": true
  },
  "globals": {
    "describe": false,
    "it": false
  },
  "rules": {
    "arrow-parens": [
      2,
      "as-needed"
    ]
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

<!-- Paste the source code below: -->
```js
const foo = async(bar) => {
//...
}

```

**What did you expect to happen?**
fix to 
```js
const foo = async bar => {
//...
}```
(or just skip the line, would be fine too)

**What actually happened? Please include the actual, raw output from ESLint.**
```js
const foo = asyncbar => {
//...
}
```

as the async was no longer `async` (but `asyncbar`), causing the function to no longer be an async function
```
/path/to/foo.js
  24:25  error  Parsing error: await is a reserved word
```

--- a/lib/rules/arrow-parens.js
+++ b/lib/rules/arrow-parens.js
@@ -51,13 +51,32 @@ module.exports = {
         const sourceCode = context.getSourceCode();
 
 
+
         /**
          * Determines whether a arrow function argument end with `)`
          * @param {ASTNode} node The arrow function node.
          * @returns {void}
          */
         function parens(node) {
-            const token = sourceCode.getFirstToken(node, node.async ? 1 : 0);
+            const isAsync = node.async;
+            const token = sourceCode.getFirstToken(node, isAsync ? 1 : 0);
+
+            /**
+             * Remove the parenthesis around a parameter
+             * @param {Fixer} fixer Fixer
+             * @returns {string} fixed parameter
+             */
+            function fixParamsWithParenthesis(fixer) {
+                const paramToken = context.getTokenAfter(token);
+                const closingParenToken = context.getTokenAfter(paramToken);
+                const asyncToken = isAsync ? context.getTokenBefore(token) : null;
+                const shouldAddSpaceForAsync = asyncToken && (asyncToken.end === token.start);
+
+                return fixer.replaceTextRange([
+                    token.range[0],
+                    closingParenToken.range[1]
+                ], `${shouldAddSpaceForAsync ? " " : ""}${paramToken.value}`);
+            }
 
             // "as-needed", { "requireForBlockBody": true }: x => x
             if (
@@ -72,15 +91,7 @@ module.exports = {
                     context.report({
                         node,
                         message: requireForBlockBodyMessage,
-                        fix(fixer) {
-                            const paramToken = context.getTokenAfter(token);
-                            const closingParenToken = context.getTokenAfter(paramToken);
-
-                            return fixer.replaceTextRange([
-                                token.range[0],
-                                closingParenToken.range[1]
-                            ], paramToken.value);
-                        }
+                        fix: fixParamsWithParenthesis
                     });
                 }
                 return;
@@ -113,15 +124,7 @@ module.exports = {
                     context.report({
                         node,
                         message: asNeededMessage,
-                        fix(fixer) {
-                            const paramToken = context.getTokenAfter(token);
-                            const closingParenToken = context.getTokenAfter(paramToken);
-
-                            return fixer.replaceTextRange([
-                                token.range[0],
-                                closingParenToken.range[1]
-                            ], paramToken.value);
-                        }
+                        fix: fixParamsWithParenthesis
                     });
                 }
                 return;

--- a/lib/rules/arrow-parens.js
+++ b/lib/rules/arrow-parens.js
@@ -50,8 +50,6 @@ module.exports = {
 
         const sourceCode = context.getSourceCode();
 
-
-
         /**
          * Determines whether a arrow function argument end with `)`
          * @param {ASTNode} node The arrow function node.

--- a/lib/rules/arrow-parens.js
+++ b/lib/rules/arrow-parens.js
@@ -59,7 +59,7 @@ module.exports = {
          */
         function parens(node) {
             const isAsync = node.async;
-            const token = sourceCode.getFirstToken(node, isAsync ? 1 : 0);
+            const firstTokenOfParam = sourceCode.getFirstToken(node, isAsync ? 1 : 0);
 
             /**
              * Remove the parenthesis around a parameter
@@ -67,13 +67,13 @@ module.exports = {
              * @returns {string} fixed parameter
              */
             function fixParamsWithParenthesis(fixer) {
-                const paramToken = context.getTokenAfter(token);
-                const closingParenToken = context.getTokenAfter(paramToken);
-                const asyncToken = isAsync ? context.getTokenBefore(token) : null;
-                const shouldAddSpaceForAsync = asyncToken && (asyncToken.end === token.start);
+                const paramToken = sourceCode.getTokenAfter(firstTokenOfParam);
+                const closingParenToken = sourceCode.getTokenAfter(paramToken);
+                const asyncToken = isAsync ? sourceCode.getTokenBefore(firstTokenOfParam) : null;
+                const shouldAddSpaceForAsync = asyncToken && (asyncToken.end === firstTokenOfParam.start);
 
                 return fixer.replaceTextRange([
-                    token.range[0],
+                    firstTokenOfParam.range[0],
                     closingParenToken.range[1]
                 ], `${shouldAddSpaceForAsync ? " " : ""}${paramToken.value}`);
             }
@@ -87,7 +87,7 @@ module.exports = {
                 node.body.type !== "BlockStatement" &&
                 !node.returnType
             ) {
-                if (astUtils.isOpeningParenToken(token)) {
+                if (astUtils.isOpeningParenToken(firstTokenOfParam)) {
                     context.report({
                         node,
                         message: requireForBlockBodyMessage,
@@ -101,12 +101,12 @@ module.exports = {
                 requireForBlockBody &&
                 node.body.type === "BlockStatement"
             ) {
-                if (!astUtils.isOpeningParenToken(token)) {
+                if (!astUtils.isOpeningParenToken(firstTokenOfParam)) {
                     context.report({
                         node,
                         message: requireForBlockBodyNoParensMessage,
                         fix(fixer) {
-                            return fixer.replaceText(token, `(${token.value})`);
+                            return fixer.replaceText(firstTokenOfParam, `(${firstTokenOfParam.value})`);
                         }
                     });
                 }
@@ -120,7 +120,7 @@ module.exports = {
                 !node.params[0].typeAnnotation &&
                 !node.returnType
             ) {
-                if (astUtils.isOpeningParenToken(token)) {
+                if (astUtils.isOpeningParenToken(firstTokenOfParam)) {
                     context.report({
                         node,
                         message: asNeededMessage,
@@ -130,8 +130,8 @@ module.exports = {
                 return;
             }
 
-            if (token.type === "Identifier") {
-                const after = sourceCode.getTokenAfter(token);
+            if (firstTokenOfParam.type === "Identifier") {
+                const after = sourceCode.getTokenAfter(firstTokenOfParam);
 
                 // (x) => x
                 if (after.value !== ")") {
@@ -139,7 +139,7 @@ module.exports = {
                         node,
                         message,
                         fix(fixer) {
-                            return fixer.replaceText(token, `(${token.value})`);
+                            return fixer.replaceText(firstTokenOfParam, `(${firstTokenOfParam.value})`);
                         }
                     });
                 }

--- a/tests/lib/rules/arrow-parens.js
+++ b/tests/lib/rules/arrow-parens.js
@@ -176,6 +176,18 @@ const invalid = [
             type
         }]
     },
+    {
+        code: "async(a) => a",
+        output: "async a => a",
+        options: ["as-needed"],
+        parserOptions: { ecmaVersion: 8 },
+        errors: [{
+            line: 1,
+            column: 1,
+            message: asNeededMessage,
+            type
+        }]
+    },
 
     // "as-needed", { "requireForBlockBody": true }
     {
@@ -214,6 +226,18 @@ const invalid = [
     },
     {
         code: "async (a) => a",
+        output: "async a => a",
+        options: ["as-needed", { requireForBlockBody: true }],
+        parserOptions: { ecmaVersion: 8 },
+        errors: [{
+            line: 1,
+            column: 1,
+            message: requireForBlockBodyMessage,
+            type
+        }]
+    },
+    {
+        code: "async(a) => a",
         output: "async a => a",
         options: ["as-needed", { requireForBlockBody: true }],
         parserOptions: { ecmaVersion: 8 },

